### PR TITLE
rocm5.3 ci

### PR DESCRIPTION
**Pull request by liligwu**
- Upgrade ROCm CI to 5.3.
- BF16 tests are disabled since BF16 methods are not implemented in ROCm.
- More managed memory allocation tests are skipped for ROCm.
- Add ROCm installation instructions in README.

### CODE REVIEW

**Review by shintaro-iwasaki — COMMENTED** [review]
Thanks for your contributions! Please ping me when this PR gets ready (e.g., after fixing Linter issues).

File: fbgemm_gpu/test/split_table_batched_embeddings_test.py [hunk 1]
```diff
@@ -157,31 +157,41 @@ def execute_forward_(
...
-                    )
+        # ROCm managed momory allocation is under development
+        if TEST_WITH_ROCM:
+            if use_cpu:
+                managed = [split_table_batched_embeddings_ops.EmbeddingLocation.HOST] * T
+                compute_device = split_table_batched_embeddings_ops.ComputeDevice.CPU
+            else:
+                managed = [
+                    split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE
+                ] * T
```
[the remark this answers is not in the document]
Comment (liligwu) [reply fbgemm_gpu/test/split_table_batched_embeddings_test.py]: > 

Thank you for your suggestions. The logic was changed.

**Comment by liligwu** [comment]
> Thanks for your contributions! Please ping me when this PR gets ready (e.g., after fixing Linter issues).

Hi @shintaro-iwasaki , regarding the linter issues, I can shorten the lines like https://github.com/pytorch/FBGEMM/blob/8184b8e84dfb69afb1b01712fa98ed9c81fd6f67/fbgemm_gpu/test/split_table_batched_embeddings_test.py#L161 but I don't think we can shorten the lines like https://github.com/pytorch/FBGEMM/blob/8184b8e84dfb69afb1b01712fa98ed9c81fd6f67/fbgemm_gpu/test/split_table_batched_embeddings_test.py#L170. Do you have any suggestions? Thank you.

**Comment by shintaro-iwasaki** [comment]
I might be misunderstanding your question so please correct me; I don't think Linter asks you to change a line including `MANAGED_CACHING`. 

Just in case: for example, one of the Linter errors is

```diff
--- /home/runner/work/FBGEMM/FBGEMM/fbgemm_gpu/test/split_table_batched_embeddings_test.py
+++ /home/runner/work/FBGEMM/FBGEMM/fbgemm_gpu/test/split_table_batched_embeddings_test.py
@@ -162,9 +162,7 @@
             compute_device = split_table_batched_embeddings_ops.ComputeDevice.CPU
         elif TEST_WITH_ROCM:
             # ROCm managed momory allocation is under development
-            managed = [
-                split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE
-            ] * T
+            managed = [split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE] * T
         elif use_cache:
             managed = [
                 split_table_batched_embeddings_ops.EmbeddingLocation.MANAGED_CACHING
```

It indicates that your current code is
```py
             compute_device = split_table_batched_embeddings_ops.ComputeDevice.CPU
         elif TEST_WITH_ROCM:
             # ROCm managed momory allocation is under development
             managed = [
                 split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE
             ] * T
         elif use_cache:
             managed = [
                 split_table_batched_embeddings_ops.EmbeddingLocation.MANAGED_CACHING
```
and the Linter asks you to change your code to
```py
             compute_device = split_table_batched_embeddings_ops.ComputeDevice.CPU
         elif TEST_WITH_ROCM:
             # ROCm managed momory allocation is under development
             managed = [split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE] * T
         elif use_cache:
             managed = [
                 split_table_batched_embeddings_ops.EmbeddingLocation.MANAGED_CACHING
```

**Comment by liligwu** [comment]
> I might be misunderstanding your question so please correct me; I don't think Linter asks you to change a line including `MANAGED_CACHING`.
> 
> Just in case: for example, one of the Linter errors is
> 
> ```diff
> --- /home/runner/work/FBGEMM/FBGEMM/fbgemm_gpu/test/split_table_batched_embeddings_test.py
> +++ /home/runner/work/FBGEMM/FBGEMM/fbgemm_gpu/test/split_table_batched_embeddings_test.py
> @@ -162,9 +162,7 @@
>              compute_device = split_table_batched_embeddings_ops.ComputeDevice.CPU
>          elif TEST_WITH_ROCM:
>              # ROCm managed momory allocation is under development
> -            managed = [
> -                split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE
> -            ] * T
> +            managed = [split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE] * T
>          elif use_cache:
>              managed = [
>                  split_table_batched_embeddings_ops.EmbeddingLocation.MANAGED_CACHING
> ```
> 
> It indicates that your current code is
> 
> ```python
>              compute_device = split_table_batched_embeddings_ops.ComputeDevice.CPU
>          elif TEST_WITH_ROCM:
>              # ROCm managed momory allocation is under development
>              managed = [
>                  split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE
>              ] * T
>          elif use_cache:
>              managed = [
>                  split_table_batched_embeddings_ops.EmbeddingLocation.MANAGED_CACHING
> ```
> 
> and the Linter asks you to change your code to
> 
> ```python
>              compute_device = split_table_batched_embeddings_ops.ComputeDevice.CPU
>          elif TEST_WITH_ROCM:
>              # ROCm managed momory allocation is under development
>              managed = [split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE] * T
>          elif use_cache:
>              managed = [
>                  split_table_batched_embeddings_ops.EmbeddingLocation.MANAGED_CACHING
> ```

Oops. Sorry, I use a local lint to analyze the code, and it gives me the "lines too long" warnings. 
I'll push the changes soon. Thank you.

**Comment by liligwu** [comment]
Hi @shintaro-iwasaki .  Would you please review the PR again?

**Review by shintaro-iwasaki — CHANGES_REQUESTED** [review]
Thanks! I left some comments, but overall this PR looks good to me!
Once all the comments are addressed, please let me import and merge this PR.

File: fbgemm_gpu/test/split_table_batched_embeddings_test.py [hunk 2]
```diff
@@ -1332,31 +1344,37 @@ def execute_backward_sgd_(  # noqa C901
...
-                average_D = sum(Ds) // T
-                for t, d in enumerate(Ds):
-                    managed[t] = (
-                        split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE
-                        if d < average_D
-                        else managed[t]
-                    )
+        # ROCm managed momory allocation is under development
+        if TEST_WITH_ROCM:
+            managed = [split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE] * T
```
[the remark this answers is not in the document]
Comment (liligwu) [reply fbgemm_gpu/test/split_table_batched_embeddings_test.py:1352]: The logic was changed.

File: fbgemm_gpu/test/split_table_batched_embeddings_test.py [hunk 3]
```diff
@@ -2492,19 +2516,25 @@ def execute_backward_optimizers_(  # noqa C901
                 np.random.randint(low=int(0.5 * E), high=int(2.0 * E)) for _ in range(T)
             ]
         compute_device = split_table_batched_embeddings_ops.ComputeDevice.CUDA
-        if use_cpu:
-            managed = [split_table_batched_embeddings_ops.EmbeddingLocation.HOST] * T
-            compute_device = split_table_batched_embeddings_ops.ComputeDevice.CPU
+        # ROCm managed momory allocation is under development
+        if TEST_WITH_ROCM:
+            managed = [split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE] * T
```
[the remark this answers is not in the document]
Comment (liligwu) [reply fbgemm_gpu/test/split_table_batched_embeddings_test.py:2518]: The same as above.

File: fbgemm_gpu/test/split_table_batched_embeddings_test.py [hunk 4]
```diff
@@ -1677,31 +1695,37 @@ def execute_backward_adagrad_(  # noqa C901
...
-                average_D = sum(Ds) // T
-                for t, d in enumerate(Ds):
-                    managed[t] = (
-                        split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE
-                        if d < average_D
-                        else managed[t]
-                    )
+        # ROCm managed momory allocation is under development
+        if TEST_WITH_ROCM:
+            managed = [split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE] * T
```
[the remark this answers is not in the document]
Comment (liligwu) [reply fbgemm_gpu/test/split_table_batched_embeddings_test.py:1700]: The same as above.

File: fbgemm_gpu/test/split_table_batched_embeddings_test.py [hunk 5]
```diff
@@ -1332,31 +1344,37 @@ def execute_backward_sgd_(  # noqa C901
...
-            ] * T
-            if mixed:
-                average_D = sum(Ds) // T
-                for t, d in enumerate(Ds):
-                    managed[t] = (
-                        split_table_batched_embeddings_ops.EmbeddingLocation.DEVICE
-                        if d < average_D
-                        else managed[t]
-                    )
+        # ROCm managed momory allocation is under development
```
[the remark this answers is not in the document]
Comment (liligwu) [reply fbgemm_gpu/test/split_table_batched_embeddings_test.py]: Thank you for the correction.

File: fbgemm_gpu/README.md [hunk 6]
```diff
@@ -93,6 +93,63 @@ python setup.py install -DTORCH_CUDA_ARCH_LIST="7.0;8.0"
...
+wget https://repo.radeon.com/amdgpu-install/5.3/ubuntu/focal/amdgpu-install_5.3.50300-1_all.deb
+sudo apt-get install ./amdgpu-install_5.3.50300-1_all.deb
+sudo amdgpu-install --usecase=hiplibsdk,rocm --no-dkms
+```
+MIOpen is required and needs to be installed seperately.
+```
+sudo apt-get install miopen-hip miopen-hip-dev
+```
+Anaconda is required.
+The remaining steps are the same as "in the container" section.
```
[the remark this answers is not in the document]
Comment (liligwu) [reply fbgemm_gpu/README.md]: It's removed.

File: fbgemm_gpu/README.md [hunk 7]
```diff
@@ -93,6 +93,63 @@ python setup.py install -DTORCH_CUDA_ARCH_LIST="7.0;8.0"
 cd bench
 python split_table_batched_embeddings_benchmark.py uvm
 ```
+## Build on ROCm
+
+FBGEMM_GPU supports running on AMD (ROCm) devices. A Docker container is recommand for setting up the ROCm environment. The installation on bare metal is also available. ROCm5.3 is used as an example of the installation below.
```
[the remark this answers is not in the document]
Comment (liligwu) [reply fbgemm_gpu/README.md]: It's corrected.

**Review by shintaro-iwasaki — APPROVED** [review]
Looks good to me! Thank you for your contribution!

It would take some time to merge it, but I will merge this PR soon.

**Review by brad-mengchi — APPROVED** [review]

**Comment by liligwu** [comment]
> Looks good to me! Thank you for your contribution!
> 
> It would take some time to merge it, but I will merge this PR soon.

Thank you for your help.

**Pull request closed without merging**